transport/x/grpc/outbound.go

```diff
@@ -0,0 +1,201 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package grpc
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+	"sync"
+	"time"
+
+	"github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc"
+	opentracing "github.com/opentracing/opentracing-go"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/errors"
+	internalsync "go.uber.org/yarpc/internal/sync"
+)
+
+var _ transport.UnaryOutbound = (*Outbound)(nil)
+var _ transport.OnewayOutbound = (*Outbound)(nil)
+
+// Outbound is a transport.UnaryOutbound.
+type Outbound struct {
+	once       internalsync.LifecycleOnce
+	lock       sync.Mutex
+	address    string
+	clientConn *grpc.ClientConn
+}
+
+// NewSingleOutbound returns a new Outbound for the given adrress.
+func NewSingleOutbound(address string) *Outbound {
+	return &Outbound{internalsync.Once(), sync.Mutex{}, address, nil}
+}
+
+// Start implements transport.Lifecycle#Start.
+func (o *Outbound) Start() error {
+	return o.once.Start(o.start)
+}
+
+// Stop implements transport.Lifecycle#Stop.
+func (o *Outbound) Stop() error {
+	return o.once.Stop(o.stop)
+}
+
+// IsRunning implements transport.Lifecycle#IsRunning.
+func (o *Outbound) IsRunning() bool {
+	return o.once.IsRunning()
+}
+
+// Transports implements transport.Inbound#Transports.
+func (o *Outbound) Transports() []transport.Transport {
+	return []transport.Transport{}
+}
+
+// Call implements transport.UnaryOutbound#Call.
+func (o *Outbound) Call(ctx context.Context, request *transport.Request) (*transport.Response, error) {
+	var responseBody []byte
+	responseMD := metadata.New(nil)
+	if err := o.invoke(ctx, request, &responseBody, &responseMD); err != nil {
+		return nil, err
+	}
+	responseHeaders, err := getApplicationHeaders(responseMD)
+	if err != nil {
+		return nil, err
+	}
+	return &transport.Response{
+		Body:    ioutil.NopCloser(bytes.NewBuffer(responseBody)),
+		Headers: responseHeaders,
+	}, nil
+}
+
+// CallOneway implements transport.OnewayOutbound#Call.
+func (o *Outbound) CallOneway(ctx context.Context, request *transport.Request) (transport.Ack, error) {
+	// pass in dummy responseBody so code doesn't complain
+	// probably safer than doing nil check in codec
+	var responseBody []byte
+	if err := o.invoke(ctx, request, &responseBody, nil); err != nil {
+		return nil, err
+	}
+	return time.Now(), nil
+}
+
+func (o *Outbound) invoke(
+	ctx context.Context,
+	request *transport.Request,
+	responseBody *[]byte,
+	responseMD *metadata.MD,
+) error {
+	start := time.Now()
+	md, err := requestToMetadata(request)
+	if err != nil {
+		return err
+	}
+	requestBody, err := ioutil.ReadAll(request.Body)
```

Comment:
Ugh, I have learned since I first wrote this that ReadAll is pretty inefficient.  We should have a pooled buffer of byte slices (similar to the internal iopool.Copy library) that we can reuse so we don't cause a lot of memory allocs.

Not sure if we can do anything similar with the response body since we're returning that to the upper applications.